### PR TITLE
strategy: elliott wave oscillator

### PR DESCRIPTION
**Pull request by zenixls2**
implement
https://tw.tradingview.com/script/3twfvd3k-Elliott-Wave-Oscillator-Signals-by-DGT/

with some modification using pinescript series features.
depends on #505 

changes were mostly inside pkg/strategy/ewo_dgtrd/strategy.go

### CODE REVIEW

**Comment by zenixls2** [comment]
```bash
BINANCE MATICUSDT PROFIT AND LOSS REPORT
===============================================
[0055]  INFO TRADES SINCE: 2022-03-19 22:01:59.999 +0900 JST
[0055]  INFO NUMBER OF TRADES: 53
[0055]  INFO AVERAGE COST: $ 1.63
[0055]  INFO TOTAL BUY VOLUME: 12728.12239991
[0055]  INFO TOTAL SELL VOLUME: 12695.11961341
[0055]  INFO STOCK: 23.45669472
[0055]  INFO CURRENT PRICE: $ 1.38
[0055]  INFO CURRENCY FEES:
[0055]  INFO  - MATIC: 9.54609178
[0055]  INFO  - USDT: 15.40986721
[0055]  INFO PROFIT: $ 479.35
[0055]  INFO UNREALIZED PROFIT: -$ 6.05
[0055]  INFO INITIAL BALANCES:
[0055]  INFO  USDT: 10000
[0055]  INFO FINAL BALANCES:
[0055]  INFO  MATIC: 33.0027865
[0055]  INFO  USDT: 10441.0308267
[0055]  INFO INITIAL ASSET IN USDT ~= 10000.0 USDT (1 MATIC = 1.669)
[0055]  INFO FINAL ASSET IN USDT ~= 10486.4 USDT (1 MATIC = 1.376)
REALIZED PROFIT: +479.35472735 USDT
UNREALIZED PROFIT: -6.0474216 USDT
ASSET INCREASED: +486.44266092 USDT (+4.86%)
```

**Comment by zenixls2** [comment]
added ma50 condition to check the trend

**Review by c9s — APPROVED** [review]

File: pkg/strategy/ewo_dgtrd/strategy.go [hunk 1]
```diff
@@ -0,0 +1,409 @@
+package ewo_dgtrd
```
Comment (c9s) [inline pkg/strategy/ewo_dgtrd/strategy.go:1]: go convention: avoid using underscore in the package name

File: pkg/strategy/ewo_dgtrd/strategy.go [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+
+func init() {
+	bbgo.RegisterStrategy(ID, &Strategy{})
+}
+
+type Strategy struct {
+	*bbgo.Graceful
+	//bbgo.SmartStops
+	market         types.Market
+	session        *bbgo.ExchangeSession
```
Comment (c9s) [inline pkg/strategy/ewo_dgtrd/strategy.go:28]: you can use:

```
	Market         types.Market
	Session        *bbgo.ExchangeSession
```

and these fields will be injected automatically.

File: pkg/strategy/ewo_dgtrd/strategy.go [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+		if len(toCancel) > 0 {
+			if err := orderExecutor.CancelOrders(ctx, toCancel...); err != nil {
+				log.WithError(err).Errorf("cancel order error")
+			}
+
+			s.tradeCollector.Process()
+		}
+
+		// well, only track prices on 1m
+		if kline.Interval == types.Interval1m {
```
Comment (c9s) [inline pkg/strategy/ewo_dgtrd/strategy.go:217]: invert this if ?

if != then return ?

Comment (zenixls2) [reply pkg/strategy/ewo_dgtrd/strategy.go:217]: condition as follows:
1. if interval == 1m, cancel and repost the orders to the latest low, also check the stoploss or trailingstop. return without further action.
2. if interval != 1m (== s.Interval), cancel all and track indicators for further decisions

line 313 will do the early return for case 1 but not in this area.

**Comment by zenixls2** [comment]
image

File: pkg/strategy/ewo_dgtrd/strategy.go [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+func init() {
+	bbgo.RegisterStrategy(ID, &Strategy{})
+}
+
+type Strategy struct {
+	*bbgo.Graceful
+	//bbgo.SmartStops
+	market         types.Market
+	session        *bbgo.ExchangeSession
+	tradeCollector *bbgo.TradeCollector
```
[the remark this answers is not in the document]
Comment (zenixls2) [reply pkg/strategy/ewo_dgtrd/strategy.go:40]: done

File: pkg/strategy/ewo_dgtrd/strategy.go (hunk 2 above)
Comment (zenixls2) [reply pkg/strategy/ewo_dgtrd/strategy.go]: done

File: pkg/strategy/ewo_dgtrd/strategy.go (hunk 1 above)
Comment (zenixls2) [reply pkg/strategy/ewo_dgtrd/strategy.go]: done

**Comment by zenixls2** [comment]
polygon:0x047d65Fca36C06Bdd9Fb2ae809732eC85d525a12

**Comment by zenixls2** [comment]
polygon:0x047d65Fca36C06Bdd9Fb2ae809732eC85d525a12

**Pull request merged**